### ISSUE TITLE
Solutions for deprecated TVIScreenCapturer streaming a custom UIWindow

### ISSUE THREAD

**Issue by jvmontes** [twilio/video-quickstart-swift#373]
Hi Twilio Team,

We are using TwilioVideo Framework version 2.7.0 on a client project. We recently started seeing issues running the application on an iPad Pro running iOS 12, similar to [this,](https://github.com/twilio/video-quickstart-swift/issues/303) and would like to figure out steps going forward to keep the same functionality we previously had.

We are using `TVIScreenCapturer` and I see that it’s deprecated: “Use ReplayKit with a custom TVIVideoSource instead.” We are currently using the `TVIScreenCapturer` and `TVILocalVideoTrack` to capture a custom window that we create. The main window shows a “host” experience, and we are creating a separate “participant” window to stream to our participants. We need to keep the custom window streaming to show the participant experience. When we displayed the topmost view, we were not capturing popovers that would display throughout the participant experience that we needed to show.

Do you have samples to accomplish the situation above using a custom TVIVideoSource? Is it necessary to include ReplayKit? We were looking to keep the user’s experience as consistent as possible with existing behavior, and I know ReplayKit may require permissions to be accepted.

Please let me know as soon as possible when you got this. If you need more time to give a more precise answer, please let me know. Thank you for your support, it’s been great working with your team and I look forward to working this out together!

**Comment by paynerc** [comment]
@jvmontes,

The problem with `TVIScreenCapturer`, and porting that functionality to a custom `TVIVideoSource` is that as of iOS 12, the methodology we used in the past is very under performant. Apple is recommending that you use ReplayKit to do screen capture and they have improved ReplayKit over the past few iOS releases.

Depending on what you are trying to capture, ReplayKit does give you access to the entire iOS device, but does require writing a ReplayKit extension. If you are only trying to capture the app experience, using the in app ReplayKit APIs is a good way to go. It should capture the entire application window, including any popovers on iPad.

The ReplayKit example app (https://github.com/twilio/video-quickstart-swift/tree/master/ReplayKitExample) contains sample code for both the extension and in app APIs. If you are using the in app APIs there are no permissions that need to be included in the Info.plist, but it will alert the user that your application wants to capture their screen and they need to approve that.

Let me know if you have further questions and we can work through them.

Ryan

**Comment by jvmontes** [comment]
Hi Ryan,

Thank you for your response. 

The problem with using the in-app ReplayKit API is that we are not using a standard screen capture of the key application window, we are creating a new window that removes certain controls and reformats UI for the participant experience. Does ReplayKit support this functionality? Does Twilio plan to add support to stream a custom UIView? 

Jorge

**Comment by paynerc** [comment]
@jvmontes,

@ceaglest and I were discussing this issue further. I was going to recommend that you take a look at #303 which describes the performance limitations in `TVIScreenCapturer` on iOS 12 but I see that you have also commented on that issue. 

The issues on iOS 12 are related to the image and color conversions that are now required to get an RGB image of the UI out of the backing store. It was for those performance reasons that when we added the new `TVIVideoSource` APIs that we didn't also implement a `TVIScreenVideoSource` class.

Apple's suggested approach for capturing screen content was to move to using ReplayKit. But for your use case, ReplayKit isn't going to work. ReplayKit doesn't give us the ability to capture just a specific view or window.

At the present time, without some assistance from Apple improving the performance, we are in a pretty tricky position. Implementing and providing a `TVIScreenVideoSource` will have pretty serious performance penalties and I think the user experience would be subpar.

However, back to your original question. You were stating that you were able to use `TVIScreenCapturer` but it wasn't capturing the popover? Are you providing the view or the window as the source for the capturer? We had an issue filed awhile back (https://github.com/twilio/twilio-video-ios/issues/3) that they said by capturing the Window they were able to get the popover to be captured.

Ryan

**Comment by phucnm** [comment]
@jvmontes AFAIK, ReplayKit always captures your main window not key window. Main window is the first one ever you set it up in the app. And it never changes. If you show UI controls in another window and make that window key and visible, RPKit won't capture them.  So it's your responsibility to handle this.

**Comment by jvmontes** [comment]
Hi,

I wanted to clarify a couple of key points just so we’re on the same page. 
@paynerc, thank you for that insight. We discovered that when we stopped sharing the main UIView and instead shared the UIWindow, we were able to see the popovers and other UI, so that isn’t an issue.
@phucnm, I want to clarify on that point, which might help explain our problem in more detail.

We have an iPad application that serves as a HOST. 
A HOST user can log in and view information for their different clients. This HOST view has navigation, controls, and several other UI components that give the HOST user power to modify information displayed. 
A PARTICIPANT is a user that is connected to a HOST and has a VIEW-ONLY experience. The HOST creates a simplified PARTICIPANT UI on a separate `UIWindow` and streams that over Twilio to remote PARTICIPANTS on a Chrome Web Browser. 
For this solution, we were using `TVIScreenCapturer`, which has unfortunately suffered performance issues since iOS 12 and Payne’s description above. The experience is very slow throughout the presentation and is essentially unusable on certain screens that are UI heavy, such as a custom animated deck experience. 

Given this, here are the paths we can see moving forward to find a solution that works.

1. Mitigate performance with the current `TVIScreenCapturer`.
    1. Use our current approach and work with Apple to mitigate the performance issues. Is there a way to reduce the amount of data we are capturing to send to the `TVIScreenCapturer`? Can we mitigate the performance issues? This approach seems risky, given that future iterations of iOS may cause further problems we can’t foresee.  
2. Work to create a solution where ReplayKit can accept a custom window.
    1. We would be able to have the same experience with some changes to our implementation of the streaming experience, with no impact to the user’s experience. 
3. Use ReplayKit and swap out the UIWindow on startup
    1. Given @phucnm's insight, would it be possible to show the participant window as our main window on startup, then switch for another window and make it key and visible? Then we would have a reference to the initial main window which could be the participant window described above.
4. Work to integrate the ReplayKit solution by requiring an additional iPad
    1. The window we display to stream over the Twilio room currently shows as the main window on a participant experience on the iPad. This solution would require additional work on all fronts, from design, development, and training to our existing users since we’ve been in a beta release. 

@paynerc and @phucnm, I would love your insight on these solutions. Do you have any connections with Apple or have any insight on the feasibility of these? 

Through the changes to this API, we are experiencing reduced functionality on our existing solution, which may put our whole release at risk. Are you aware of other clients that are using the screen sharing functionality in this way? Is this a unique use case?

Thank you for your continued support as we work through this!

**Comment by phucnm** [comment]
@jvmontes  https://developer.apple.com/videos/play/wwdc2016/601/?time=516. They talked about how to show an Indicator View on a new window. It's stated in Transcripts as well. I couldn't find in official docs in I has been working with RPKit for a while. Hope this help you decide your view hierarchy.

**Comment by paynerc** [comment]
@jvmontes,

After also watching the session that @phucnm suggested, it would seem that it should work making the participant window be the initial window at app launch and then quickly adding your host window. The assumption there that when you start broadcasting through ReplayKit it will go back to that initial participant window.

Let us know if that solution does indeed work for you.

Ryan

**Comment by jvmontes** [comment]
Hey all,

Sorry for the delayed response. I've been working on getting this fix in, but yes, it appears that we are able to show an alternate screen on the receiving end of the Twilio stream on a Chrome Browser from our iPad. Now we are just working on making sure the participant screen follows along with the host navigation.. essentially rewiring up all the components. 

Thank you @paynerc @phucnm for your support to cross this hurdle, we've made some great progress!! 😃

**Comment by paynerc** [comment]
@jvmontes,

That is fantastic news! Do you think we are good to close this issue as the workaround seems to be functioning?

Let us know how things progress.

Ryan

**Comment by jvmontes** [comment]
Here's the proof of concept, for anyone else that may be facing this issue:

We had our `Main.storyboard` as our host experience, and a `Participant.storyboard` as the participant experience. The host experience is what we want displayed on the iPad, the participant experience is what we want streamed on Twilio. 

To make our app work with the way ReplayKit streams meetings, we switched the participant to be `Main.storyboard` and renamed our host storyboard to `Host.storyboard`. 

In the `AppDelegate.swift`, we added 
`fileprivate var hostWindow: UIWindow?`
and in `func applicationDidBecomeActive()`, we added the following code:
```       
        if hostWindow == nil {
            hostWindow = UIWindow.init(frame: UIScreen.main.bounds)
            hostWindow?.makeKeyAndVisible()
            let hostStoryboard = UIStoryboard(name: "Host", bundle: nil)
            let hostController = hostStoryboard.instantiateInitialViewController()
            hostWindow?.rootViewController = hostController
        }
```

When we initiate our ReplayKit stream, the window that was initially shown is being streamed. 

Good to close!
